### PR TITLE
maven plugin: add support for bases

### PR DESCRIPTION
**Pull request by sergiusens**
Update the Maven plugin to be base-aware, only supporting core16 and
core18.

LP: #1794793

Signed-off-by: Sergio Schvezov <sergio.schvezov@canonical.com>

- [ ] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [ ] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [ ] If this is a bugfix. Have you checked that there is a bug report open for the issue you are trying to fix on [bug reports](https://bugs.launchpad.net/snapcraft)?
- [ ] If this is a new feature. Have you discussed the design on the [forum](https://forum.snapcraft.io)?
- [ ] Have you successfully run `./runtests.sh static`?
- [ ] Have you successfully run `./runtests.sh tests/unit`?

-----

### CODE REVIEW

File: snapcraft/internal/sources/_base.py [hunk 1]
```diff
@@ -124,21 +124,21 @@ def pull(self):
...
         file_cache = FileCache()
         if self.source_checksum:
             algorithm, hash = split_checksum(self.source_checksum)
             cache_file = file_cache.get(algorithm=algorithm, hash=hash)
             if cache_file:
-                self.file = os.path.join(self.source_dir, os.path.basename(cache_file))
                 # We make this copy as the provisioning logic can delete
                 # this file and we don't want that.
                 shutil.copy2(cache_file, self.file)
+                print(self.file)
```
Comment (kyrofa) [inline snapcraft/internal/sources/_base.py]: Leftover debugging?

File: snapcraft/plugins/maven.py [hunk 2]
```diff
@@ -64,39 +90,108 @@ def schema(cls):
...
+        super().__init__(name, options, project)
+
+        self._setup_maven()
+        self._setup_base_tools(project.info.base)
+
+    def _setup_base_tools(self, base):
+        version = self.options.maven_openjdk_version
+        if base == "core16":
+            if not version:
+                version = "8"
```
Comment (kyrofa) [inline snapcraft/plugins/maven.py]: Why 8 and not 9?

File: snapcraft/plugins/maven.py [hunk 3]
```diff
@@ -64,39 +90,108 @@ def schema(cls):
...
+        return self._maven_tar_handle
+
+    def __init__(self, name, options, project):
+        super().__init__(name, options, project)
+
+        self._setup_maven()
+        self._setup_base_tools(project.info.base)
+
+    def _setup_base_tools(self, base):
+        version = self.options.maven_openjdk_version
```
Comment (kyrofa) [inline snapcraft/plugins/maven.py]: If the select e.g. 11 while on core16, they will get the not-so-friendly "package not found" error. Do you think that's clear enough, or would it be better to limit the possibilities to what we know is actually possible?

File: tests/spread/plugins/maven/legacy-build-and-run/task.yaml [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+restore: |
+  cd "$SNAP_DIR"
+  snapcraft clean
+  rm -f ./*.snap
+
+execute: |
+  cd "$SNAP_DIR"
+  snapcraft
+  sudo snap install maven-hello_*.snap --dangerous
+  [ "$(maven-hello)" = "Hello World!" ]
```
Comment (kyrofa) [inline tests/spread/plugins/maven/legacy-build-and-run/task.yaml]: We've sort of established two different patterns here: legacy pulling, and legacy running (i.e. entire lifecycle and install/run). Is there something gained by doing more than pulling, here?

File: snapcraft/plugins/maven.py (hunk 3 above)
Comment (sergiusens) [reply snapcraft/plugins/maven.py]: yes, I thought I had, good catch

File: snapcraft/plugins/maven.py (hunk 2 above)
Comment (sergiusens) [reply snapcraft/plugins/maven.py]: I was trying something out, let me move that out of the way.

File: tests/spread/plugins/maven/legacy-build-and-run/task.yaml (hunk 4 above)
Comment (sergiusens) [reply tests/spread/plugins/maven/legacy-build-and-run/task.yaml]: There is nothing and the original tests actually do not even try and run.

Comment (sergiusens) [reply tests/spread/plugins/maven/legacy-build-and-run/task.yaml]: I think that is what you adopted versus what I adopted
```
tests/spread/plugins/dotnet/legacy-build-run
tests/spread/plugins/waf/legacy-build-run
tests/spread/plugins/ruby/legacy-pull
tests/spread/plugins/maven/snaps/legacy-maven-hello
tests/spread/plugins/maven/legacy-build-and-run
tests/spread/plugins/go/legacy-build-run
tests/spread/plugins/plainbox/legacy-build-run
tests/spread/plugins/meson/snaps/legacy-meson
tests/spread/plugins/meson/legacy-build-run
tests/spread/plugins/autotools/legacy-pull
tests/spread/plugins/catkin/legacy-pull
tests/spread/plugins/ament/legacy-pull
tests/spread/plugins/make/legacy-pull
tests/spread/plugins/scons/legacy-build-run
tests/spread/plugins/cmake/legacy-pull
tests/spread/plugins/godeps/legacy-build-run
```
With the exception of ament where I admit I just did pull to save time.

**Pull request merged**